new: introduce `is_exe_upper_layer`

loresuso opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area driver-kmod

/area driver-ebpf

/area libscap

/area libsinsp

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:

This PR tries to deal with `overlayfs`, a union mount filesystem often used by container runtimes, to understand if a process is executing something that is part of the base image of a container or not. Monitoring this could be relevant from a security standpoint since best practices state that containers should be immutable. For this reason, in many use cases, we don't expect processes in the container to execute something that is not part of the base image, whether that file was created in the container or an existing one was modified. 
To do so, the PR aims at identifying the layer of the file being executed by a process: if the file belongs to the lower layer, it means that it was part of the base image of the container. Otherwise, the process executed something that is part of the upper layer, and a flag is set upon the `execve` event.

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

To better understand the code, I'd like to give reviewers some useful pointers. First, we must check if the file being executed is part of an overlayfs. We can do that by retrieving the superblock magic of the file and comparing it with the `OVERLAYFS_SUPER_MAGIC`.  You can see [here](https://elixir.bootlin.com/linux/latest/source/fs/overlayfs/super.c#L2129) that this constant is used when an overlayfs is mounted. 
Then, to check the layer of the executable file, we can do some pointer arithmetics, having in mind that the pointer to the inode that we can retrieve form `struct file` is pointing to the `vfs_inode` member of the `ovl_inode` struct, as you can see [here](https://elixir.bootlin.com/linux/latest/source/fs/overlayfs/ovl_entry.h#L122).

The last thing that I wanted to discuss is that for now I didn't find a way to retrieve the upper_layer information when scanning the proc filesystem. One possible way to do it could be by inspecting the `/proc/<pid>/mounts` file and checking if there is an overlayfs entry, which in case we can use to retrieve the upperdir directory and look for the `exe` file there. However, this would easily work when Falco is installed on the host, but could be a problem when it is running as a container. This is because most of the time the upperdir directory is placed in `/var/lib/docker`, `/var/lib/containerd` directories. This means we would need to give the Falco container access to those directories. I would like to know what do you think, and if you like it, I can implement it. Also, I am open to any other possible solution on this!

EDIT: I have implemented this approach

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
new: introduce is_exe_upper_layer
```

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/287#" title="Author self-approved">loresuso</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **fntlnz** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @fntlnz` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["fntlnz"]} -->

poiana commented:
Hi @loresuso. Thanks for your PR.

I'm waiting for a [falcosecurity](https://github.com/orgs/falcosecurity/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Flibs).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

FedeDP commented:
This is really cool @loresuso ! Thanks for this effort!

FedeDP commented:
/ok-to-test

FedeDP reviewed (commented):
Changes LGTM; my only concern (and the same goes for exe_writable for example) is that we are returning "false" by default where unsupported (kernel versions < 3.18 using kmod), while we do not have any info really.
I don't think it is a major issue, but i don't really like returning wrong informations to clients.

loresuso commented on userspace/libsinsp/threadinfo.cpp:
```diff
@@ -1544,7 +1545,8 @@ void sinsp_thread_manager::dump_threads_to_file(scap_dumper_t* dumper)
 			sizeof(uint64_t) + // cap_permitted
 			sizeof(uint64_t) + //cap_effective
-			sizeof(uint8_t); // exe_writable
+			sizeof(uint8_t) + // exe_writable
```
Of course we can! Updated!

loresuso commented:
Hey @FedeDP, thank you! 
I agree with you, maybe we can better document the fact those new fields (`is_exe_writable` and `is_exe_upper_layer`) can be trusted only on specific kernel versions? I thought about that too, didn't know how to solve it honestly

FedeDP commented:
> I agree with you, maybe we can better document the fact those new fields (is_exe_writable and is_exe_upper_layer) can be trusted only on specific kernel versions? I thought about that too, didn't know how to solve it honestly

Yeah that seems a reasonable thing to do.  
Another way of solving would be to use a string instead of a boolean and returning <NA> where the info is not available, but i don't quite like the idea :)

FedeDP commented:
> Yeah, I don't like that too, execve gets called pretty much of course, let's not waste space on the ring buffer just for this :)

I meant that we can push onto the ring buffer eg: [ -1, 0, 1] where -1 means "NA", 0 means "false" and 1 means "true". Ring buffer would be untouched.
What i don't really like is having the filterchecks on a string instead of a boolean, it is much less elegant.

loresuso commented:
Ah I got it now, sorry! Yeah you are right

FedeDP commented on driver/bpf/fillers.h:
```diff
@@ -2083,6 +2083,39 @@ static __always_inline bool get_exe_writable(struct task_struct *task)
+	}
+
+	struct inode *exe_inode = _READ(exe_file->f_inode);
+	struct super_block *sb = _READ(exe_inode->i_sb);
```
I see that basically until this loc, this functions shares 100% of the code with `get_exe_writable`.
I am not sure; is it costly to call all of these?
Can't we have a single function to extract both booleans?

loresuso commented:
Hey @FedeDP, I think I have addressed all your comments:
- Stated that `is_exe_upper_layer` is meaningful if underlying kernel version is above 3.18
- Added a small function `get_exe_inode` to retrieve the inode of the executable. It is now used by `exe_writable` and `exe_upper_layer`. 
-  decided to add a function to properly set the `upper_layer` flag when scanning `procfs`. I tested it with docker and containerd and it seems perfectly ok, let me know what do you think on this
- Noted that  the overlays magic constant was defined in a c file before kernel 4.5, so to make it work, I defined PPM_OVERLAYFS_SUPER_MAGIC as you can see

Thank you! Your feedback is always much appreciated!!

LucaGuerra commented on userspace/libscap/scap_procs.c:
```diff
@@ -700,6 +700,83 @@ int32_t scap_proc_fill_exe_writable(scap_t* handle, struct scap_threadinfo* tinf
+		start = strstr(line, "upperdir=") + sizeof("upperdir=") - 1;
+		if(!start)
+		{
+			return;
```
Seems like some cases like this do not `fclose()` the file

loresuso replied:
Thanks for noticing this!

LucaGuerra commented on userspace/libscap/scap_procs.c:
```diff
@@ -700,6 +700,83 @@ int32_t scap_proc_fill_exe_writable(scap_t* handle, struct scap_threadinfo* tinf
+	// and set upper_layer accordingly
+	//
+	strncat(upperdir, exe, sizeof(upperdir) - strlen(upperdir) - 1);
+	ret = stat(upperdir, &upperdir_exe_stat);
```
If I understand the description right, that would be the part that fails in case Falco runs in a container. I believe it would need the right host prefix to work in that case if the appropriate directory is mounted. The other workaround would be to use `/proc/1/root/...` but I agree it's not the cleanest way of doing that.

loresuso replied:
I found out the function `scap_get_host_root` and I think it solves the problem. Falco uses the environment variable `HOST_ROOT`, which is set to `/host` in the official docker image. I believe that if we mount directories such as `/var/lib/docker`, their path will follow this pattern too, ending up in `/host/var/lib/docker`. So just prepending it to each path should work, thank you! :)

loresuso replied:
I am going to try it!

loresuso replied:
I have tested it, and it now works also when Falco runs in a container if `/var/lib/docker` is mounted. Thank you!

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/287#" title="Author self-approved">loresuso</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **fededp** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @fededp` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["fededp"]} -->

loresuso commented:
/retest

poiana commented:
@loresuso: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | dcba40b40f3d7f1d17352e6e7e1717efb2ec581d | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/287/build-libs-arm/1531578302194520064) | true | `/test build-libs-arm`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=287). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@loresuso: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | dcba40b40f3d7f1d17352e6e7e1717efb2ec581d | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/287/build-libs-arm/1531578302194520064) | true | `/test build-libs-arm`
build-libs | dcba40b40f3d7f1d17352e6e7e1717efb2ec581d | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/287/build-libs/1531578302102245376) | true | `/test build-libs`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=287). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@loresuso: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | dcba40b40f3d7f1d17352e6e7e1717efb2ec581d | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/287/build-libs-arm/1531578302194520064) | true | `/test build-libs-arm`
build-libs | dcba40b40f3d7f1d17352e6e7e1717efb2ec581d | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/287/build-libs/1531578302102245376) | true | `/test build-libs`
build-libs-bundled-deps | dcba40b40f3d7f1d17352e6e7e1717efb2ec581d | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/287/build-libs-bundled-deps/1531578302249046016) | true | `/test build-libs-bundled-deps`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=287). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@loresuso: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | dcba40b40f3d7f1d17352e6e7e1717efb2ec581d | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/287/build-libs-arm/1531578302194520064) | true | `/test build-libs-arm`
build-libs | dcba40b40f3d7f1d17352e6e7e1717efb2ec581d | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/287/build-libs/1531578302102245376) | true | `/test build-libs`
build-libs-bundled-deps | dcba40b40f3d7f1d17352e6e7e1717efb2ec581d | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/287/build-libs-bundled-deps/1531578302249046016) | true | `/test build-libs-bundled-deps`
build-libs-with-chisels | dcba40b40f3d7f1d17352e6e7e1717efb2ec581d | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/287/build-libs-with-chisels/1531578302341320704) | true | `/test build-libs-with-chisels`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=287). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@loresuso: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | b8cdc809a553467a34e27b58f83caec507f976e0 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/287/build-libs-arm/1545428047128170496) | true | `/test build-libs-arm`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=287). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

loresuso commented:
Hey maintainers! I have rebased this PR again, and I still think is a good feature to add. What do you think? 🙂

loresuso commented:
/cc @FedeDP @Andreagit97 @LucaGuerra

poiana commented:
@loresuso: GitHub didn't allow me to request PR reviews from the following users: LucaGuerra.

Note that only [falcosecurity members](https://github.com/orgs/falcosecurity/people) and repo collaborators can review this PR, and authors cannot review their own PRs.

<details>

In response to [this](https://github.com/falcosecurity/libs/pull/287#issuecomment-1192686090):

>/cc @FedeDP @Andreagit97 @LucaGuerra 


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

FedeDP commented on driver/bpf/fillers.h:
```diff
@@ -2248,6 +2256,29 @@ static __always_inline bool get_exe_writable(struct task_struct *task)
+		bpf_probe_read(&upper_dentry, sizeof(upper_dentry), vfs_inode + sizeof(struct inode));
+		bpf_probe_read(&lower_inode, sizeof(lower_inode), vfs_inode + sizeof(struct inode) + sizeof(struct dentry *));
+
+		if(!lower_inode && upper_dentry)
```
I'd pre-initialize both to NULL so that even if `bpf_probe_read` fails for some reason, we still have a coherent behavior.

Moreover, i'm quite wary about adding the pointer arithmetic above :/ You know, that weird sense of fear with eBPF :D

FedeDP replied:
I'd also double check the minimum kernel requirements to see if that holds true on any kernel supported (ie: >= 4.14).

FedeDP commented on driver/bpf/fillers.h:
```diff
@@ -2248,6 +2256,29 @@ static __always_inline bool get_exe_writable(struct task_struct *task)
 
+static __always_inline bool get_exe_upper_layer(struct inode *inode)
+{
+	struct super_block *sb = _READ(inode->i_sb);
```
What about checking if `sb != NULL` here?
Same in the above usage inside `get_exe_writable` :)

FedeDP commented on driver/bpf/fillers.h:
```diff
@@ -6252,7 +6298,8 @@ FILLER(sched_prog_exec_4, false)
 	/* `exe_writable` flag logic */
 	bool exe_writable = false;
-	exe_writable = get_exe_writable(task);
+	struct inode *inode = get_exe_inode(task);
```
In `execve_family_flags` filler we actually check that `inode != NULL` before using it. Shall we do the same here?

FedeDP replied:
Moreover i think you miss the `exe_upper_layer` in the `sched_prog_exec_4` filler (that is the new filler for exec for arm64 and s390x).

FedeDP commented on driver/ppm_fillers.c:
```diff
@@ -1269,16 +1270,43 @@ int f_proc_startupdate(struct event_filler_arguments *args)
+					struct super_block *sb;
+					unsigned long sb_magic;
+
+					sb = exe_file->f_inode->i_sb;
```
As for the eBPF probe, what about checking if `sb != NULL` ?

FedeDP commented on userspace/libscap/scap_procs.c:
```diff
@@ -665,6 +665,86 @@ int32_t scap_proc_fill_exe_writable(scap_t* handle, struct scap_threadinfo* tinf
+	// and we can retrieve the upperdir path for this container.
+	//
+	if(!fgets(line, sizeof(line), mounts_file))
+	{
```
`fclose(mounts_file)` missing!

FedeDP commented on driver/ppm_fillers.c:
```diff
@@ -1269,16 +1270,43 @@ int f_proc_startupdate(struct event_filler_arguments *args)
 			flags |= PPM_EXE_WRITABLE;
 		}
 
+		if (exe_upper_layer) {
```
This is ok.
Like for the eBPF probe, i think you miss the `exe_upper_layer` in the `f_sched_prog_exec` filler (that is the new filler for exec for arm64 and s390x).

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/287#" title="Author self-approved">loresuso</a>*
**Once this PR has been reviewed and has the lgtm label**, please ask for approval from fededp by writing `/assign @fededp` in a comment. For more information see:[The Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["fededp"]} -->

leogr commented:
/milestone 0.10.0

incertum commented on driver/bpf/fillers.h:
```diff
@@ -2239,6 +2251,33 @@ static __always_inline bool get_exe_writable(struct task_struct *task)
+		bpf_probe_read(&upper_dentry, sizeof(upper_dentry), vfs_inode + sizeof(struct inode));
+		bpf_probe_read(&lower_inode, sizeof(lower_inode), vfs_inode + sizeof(struct inode) + sizeof(struct dentry *));
+
+		if(!lower_inode && upper_dentry)
```
Curiosity question: This double check condition is not redundant right? Asking because we don't seem to care about the actual fields and rather only use them to set the boolean value if there is an `upper_dentry`. Is it there is only ever either or?

 `bpf_fd_to_socktuple` seems to do similar pointer arithmetic.

incertum reviewed (commented):
Thank you @loresuso for this :rocket: PR! Left only a few nit comments.
In addition, compiled and tested the actual new capability. Is working as expected :smile:, nice job!

incertum commented:
One more thought/question: You mentioned that this does not work for all runtimes, would it make sense to mention the runtimes that are compatible in the definitions?

incertum commented:
@loresuso as `modern_bpf` is one the near-term roadmap for initial experimental release, would it be possible to already add this new signal to `modern_bpf` as well? Would be much appreciated. CC @Andreagit97

CC @LucaGuerra would also appreciate adding the exe writable signal to `modern_bpf` (if possible) to have feature parity across all 3 drivers.

Andreagit97 commented on driver/bpf/fillers.h:
```diff
@@ -2151,20 +2151,32 @@ static __always_inline bool bpf_kgid_has_mapping(struct user_namespace *targ, kg
 	struct file *exe_file;
 	struct mm_struct *mm;
+	
 	mm = _READ(task->mm);
```
Agree with @incertum here we need to choose a way to handle this `NULL` check. To be honest, I would avoid the check unless it is necessary or explicitly required by the verifier... every check has a ridiculous cost but if we put many of them in the same flow it could impact perf.

Andreagit97 replied:
and adding them we increase also the bytecode complexity with many jumps

Andreagit97 replied:
Until we use the pointer only to perform another `_READ` we should be safe without the `NULL` check, this is the same logic I've followed in the modern probe, obviously here we are talking about lower kernel versions so never say never :roll_eyes:

Andreagit97 commented on driver/bpf/fillers.h:
```diff
@@ -2151,20 +2151,32 @@ static __always_inline bool bpf_kgid_has_mapping(struct user_namespace *targ, kg
 
 	struct inode *inode = _READ(exe_file->f_inode);
+
+	return inode;
```
`NULL` should be checked outside otherwise the verifier could go crazy with some clang/kernel versions, so yes if the verifier allows us to do it, we can return like this `return _READ(exe_file->f_inode);`

Andreagit97 commented on driver/bpf/fillers.h:
```diff
@@ -2151,20 +2151,32 @@ static __always_inline bool bpf_kgid_has_mapping(struct user_namespace *targ, kg
+	return inode;
+}
+
+static __always_inline bool get_exe_writable(struct task_struct *task, struct inode *inode)
```
yeah agree also on that we have just to understand in which order we want to merge them

incertum replied:
We can merge this first if ok with @loresuso since it has been open for a while, happy to rebase the other PR then.

Andreagit97 commented on driver/bpf/fillers.h:
```diff
@@ -2239,6 +2251,33 @@ static __always_inline bool get_exe_writable(struct task_struct *task)
+		char *vfs_inode = (char *)inode;
+		
+		bpf_probe_read(&upper_dentry, sizeof(upper_dentry), vfs_inode + sizeof(struct inode));
+		bpf_probe_read(&lower_inode, sizeof(lower_inode), vfs_inode + sizeof(struct inode) + sizeof(struct dentry *));
```
Pain point here :/ the `struct ovl_inode` structure is already changed:
-> 4.14 https://elixir.bootlin.com/linux/v4.14/source/fs/overlayfs/ovl_entry.h#L60
-> 6.0.9 https://elixir.bootlin.com/linux/latest/source/fs/overlayfs/ovl_entry.h#L127

`struct inode *lower_inode` doesn't exist anymore :/

Andreagit97 replied:
it changes in kernel version `5.19`

incertum replied:
See previous remark https://github.com/falcosecurity/libs/pull/287#discussion_r973520058

Let's just get rid of the redundant check :) tested in on a 6.0.8 kernel, works:

```
unsigned long sb_magic = _READ(sb->s_magic);
if(sb_magic == PPM_OVERLAYFS_SUPER_MAGIC)
{
	struct dentry *upper_dentry = NULL;
	char *vfs_inode = (char *)inode;
	bpf_probe_read(&upper_dentry, sizeof(upper_dentry), vfs_inode + sizeof(struct inode));
	if(upper_dentry)
	{
		return true;
	}
}
```

Andreagit97 replied:
it would be amazing!!

loresuso replied:
Yes, it seems reasonable! Great Melissa, I think this is the best way to check this.
Updated with this change

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/287#" title="Author self-approved">loresuso</a>*
**Once this PR has been reviewed and has the lgtm label**, please ask for approval from fededp by writing `/assign @fededp` in a comment. For more information see [the Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["fededp"]} -->

loresuso commented:
Hello everybody, I have reworked this PR addressing the comments and implemented the related version in the modern eBPF probe, so that we have feature parity in all three instrumentation modes.
I also tested the old eBPF probe and the kmod on older kernels, like version 4.18 and unfortunately we are not able to retrieve the upper layer information. It seems that the kernel uses a `sb_magic` that is different in newer kernel and associates it to ext3 in my experiments.  

If everyone like the approach of this PR, I can work on finding the minimum version from which it starts to be working and add a comment about it inside the code, but first I would like to hear everyone's opinion since this is a really time consuming task.

Andreagit97 commented on driver/bpf/fillers.h:
```diff
@@ -6416,13 +6449,24 @@ FILLER(sched_prog_exec_4, false)
+	if(inode)
 	{
-		flags |= PPM_EXE_WRITABLE;
+		exe_writable = get_exe_writable(task, cred);
```
```suggestion
		exe_writable = get_exe_writable(inode, cred);
```

loresuso replied:
Thanks a lot!

Andreagit97 commented:
This PR LGTM! Thank you for all the effort you put into it! :tada: 

> If everyone like the approach of this PR, I can work on finding the minimum version from which it starts to be working and add a comment about it inside the code, but first I would like to hear everyone's opinion since this is a really time consuming task.

Yeah, it would be amazing to find the first working version!

> Also, I am thinking about ways to introduce tests for this feature. A possible way would be to mount an overlayfs and try to execute something from there and verify that the flag gets set.

Absolutely, I think we need a test, in this way we can easily detect when this access `vfs_inode + sizeof(struct inode));` is no more valid

Andreagit97 reviewed (approved):
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: 54081517502da7292c059a57b9aff84e49779ba5</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/287#pullrequestreview-1194518560" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/287#" title="Author self-approved">loresuso</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/287#pullrequestreview-1194518560" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/287#pullrequestreview-1194521013" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/287#" title="Author self-approved">loresuso</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
